src/core/vector/qgsvectorlayerprofilegenerator.cpp

```diff
@@ -721,6 +722,83 @@ bool QgsVectorLayerProfileGenerator::generateProfile( const QgsProfileGeneration
   if ( !mProfileCurve || mFeedback->isCanceled() )
     return false;
 
+  auto profileLine = qgsgeometry_cast<const QgsLineString *>( mProfileCurve.get() );
+  if ( profileLine )
+  {
+    // The profile generation code can't deal with curves that enter a single
+    // point multiple times. We handle this for line strings by splitting them
+    // into multiple parts, each with no repeated points, and computing the
+    // profile for each by itself.
+    std::unique_ptr< QgsVectorLayerProfileResults > totalResults;
+    double distanceProcessed = 0;
+
+    auto processPart = [this, &totalResults, &distanceProcessed]( QgsPointSequence & points )
+    {
+      mProfileCurve.reset( new QgsLineString( points ) );
+      if ( !generateProfileInner() ) return false;
+
+      if ( !totalResults )
+        // Use the first result set as a base
+        totalResults.reset( mResults.release() );
+      else
+      {
+        // Merge the results, shifting them by distanceProcessed
+        totalResults->mRawPoints.append( mResults->mRawPoints );
+        totalResults->minZ = std::min( totalResults->minZ, mResults->minZ );
+        totalResults->maxZ = std::max( totalResults->maxZ, mResults->maxZ );
+        for ( auto it = mResults->mDistanceToHeightMap.constKeyValueBegin();
+              it != mResults->mDistanceToHeightMap.constKeyValueEnd();
+              ++it )
+        {
+          totalResults->mDistanceToHeightMap[it->first + distanceProcessed] = it->second;
+        }
+        for ( auto it = mResults->features.constKeyValueBegin();
+              it != mResults->features.constKeyValueEnd();
+              ++it )
+        {
+          for ( QgsVectorLayerProfileResults::Feature feature : it->second )
+          {
+            feature.crossSectionGeometry.translate( distanceProcessed, 0 );
+            totalResults->features[it->first].push_back( feature );
+          }
+        }
+      }
+
+      distanceProcessed += mProfileCurve->length();
+      return true;
+    };
+
+    QgsPointSequence allPoints;
+    profileLine->points( allPoints );
```

Comment:
This is an inefficient approach to iterate through the points in a linestring -- see eg https://github.com/qgis/QGIS/blob/master/src/core/symbology/qgssymbollayerutils.cpp#L4452 . for the fastest approach

Reply:
Thanks, I've switched to iterating the data directly.

Is there a reason that there's no iterator of `QgsPoint` on `QgsLineString`?

Reply:
> Is there a reason that there's no iterator of QgsPoint on QgsLineString?

They aren't stored natively as QgsPoint in QgsLineString, so any iterator would be constructing a bunch of new objects and would be inefficient as a result. That's why it's recommended to iterate the co-ordinate arrays directly.

On this note it's also MUCH faster to construct QgsLineString from Qvector<double> containers for X/y, rather then the QgsPointSequence approach used here.

Reply:
> They aren't stored natively as QgsPoint in QgsLineString, so any iterator would be constructing a bunch of new objects and would be inefficient as a result. That's why it's recommended to iterate the co-ordinate arrays directly.

I would hope that the temporary object construction would be optimised by the compiler, but I've never checked. In this specific case it would lead to much nicer code, since I'm constructing `QgsPoint`s anyway.

> On this note it's also MUCH faster to construct QgsLineString from Qvector containers for X/y, rather then the QgsPointSequence approach used here.

I need a single object to put into the `QSet` for duplicate detection, so I can't avoid `QgsPoint`s entirely, but I'll change the method to collect two separate arrays

Reply:
> need a single object to put into the QSet for duplicate detection

You possibly want QgsPointXY for that, which will take half the storage space if z/m aren't important

Reply:
> You possibly want QgsPointXY for that, which will take half the storage space if z/m aren't important

Thanks, I've switched the set to `QgsPointXY`. I've also renamed the method, since a few changes back it started throwing away non-XY coordinates. That's what I want for the elevation profile, but I don't want to surprise people expecting a more general method.